hledger-lib/Hledger/Data/DayPartition.hs

```diff
@@ -0,0 +1,275 @@
+{-|
+A partition of time into contiguous spans, for defining reporting periods.
+-}
+module Hledger.Data.DayPartition
+( DayPartition
+, boundariesToDayPartition
+, boundariesToMaybeDayPartition
+
+, lookupDayPartition
+, unionDayPartitions
+
+, dayPartitionToNonEmpty
+, dayPartitionToList
+, dayPartitionSpans
+, dayPartitionToDateSpans
+, dayPartitionToPeriodData
+, maybeDayPartitionToDateSpans
+
+, splitSpan
+, intervalBoundaryBefore
+
+, tests_DayPartition
+) where
+
+import qualified Data.IntMap.Strict as IM
+import Data.List.NonEmpty (NonEmpty(..))
+import qualified Data.List.NonEmpty as NE
+import Data.Time (Day(..), addDays, addGregorianMonthsClip, addGregorianYearsClip, fromGregorian)
+
+import Hledger.Data.Dates
+import Hledger.Data.PeriodData
+import Hledger.Data.Types
+import Hledger.Utils
+
+
+-- | A partition of time into contiguous spans, along with a historical period
+-- before any of the spans.
+--
+-- This is a newtype wrapper around 'PeriodData Day', where the start dates are
+-- the keys and the end dates are the values. Spans are stored in inclusive format
+-- [start, end]. Note that this differs from 'DateSpan' which uses [start, end)
+-- format.
```

Comment:
Thanks for expanding these descriptions. I think we can do better still; as a person reading this, not familiar with PeriodData (or even if I am), I find it not too clear. "Start dates are keys, end dates are values." I don't see it.